File: backend-go/internal/servicebill/controller/servicebill_controller.go
package controller

import (
	"backend-go/internal/common/errs"
	"backend-go/internal/common/result"
	"backend-go/internal/servicebill/model"
	"backend-go/internal/servicebill/service"
	"strconv"

	"github.com/gofiber/fiber/v2"
)

type ServiceBillController struct {
	bizSrv *service.BizService
}

func NewServiceBillController(bizSrv *service.BizService) *ServiceBillController {
	return &ServiceBillController{
		bizSrv: bizSrv,
	}
}

func (c *ServiceBillController) QueryByParam(ctx *fiber.Ctx) error {
	var param model.ServiceBillQueryParam
	if err := ctx.BodyParser(&param); err != nil {
		return err
	}

	res, err := c.bizSrv.FindByParam(ctx.Context(), &param)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) GetByID(ctx *fiber.Ctx) error {
	idStr := ctx.Params("id")
	if idStr == "" {
		return errs.NewBizError("ID 为空")
	}
	id, err := strconv.Atoi(idStr)
	if err != nil {
		return err
	}
	res, err := c.bizSrv.FindByID(ctx.Context(), uint(id))
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Create(ctx *fiber.Ctx) error {
	var dto model.ServiceBillDTO
	if err := ctx.BodyParser(&dto); err != nil {
		return err
	}

	res, err := c.bizSrv.Create(ctx.Context(), &dto)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) ImportByFile(ctx *fiber.Ctx) error {
	form, err := ctx.MultipartForm()
	if err != nil {
		return err
	}
	files := form.File["file"]
	if len(files) == 0 {
		return errs.NewBizError("没有文件上传")
	}
	res, err := c.bizSrv.GenerateByFile(ctx.Context(), files[0])
	if err != nil {
		return err
	}

	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Update(ctx *fiber.Ctx) error {
	var dto model.ServiceBillDTO
	if err := ctx.BodyParser(&dto); err != nil {
		return err
	}

	res, err := c.bizSrv.Update(ctx.Context(), &dto)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Delete(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.bizSrv.Delete(ctx.Context(), ids)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Process(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.bizSrv.Process(ctx.Context(), ids)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Processed(ctx *fiber.Ctx) error {
	var param model.ProcessedParam
	if err := ctx.BodyParser(&param); err != nil {
		return err
	}

	res, err := c.bizSrv.Processed(ctx.Context(), param.Ids, param.ProcessedDate)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Finish(ctx *fiber.Ctx) error {
	var param model.FinishParam
	if err := ctx.BodyParser(&param); err != nil {
		return err
	}

	res, err := c.bizSrv.Finish(ctx.Context(), param.Ids, param.FinishedDate)
	if err != nil {
		return err
	}
	result.SetResult(ctx, res)
	return nil
}

func (c *ServiceBillController) Export(ctx *fiber.Ctx) error {
	var ids []uint
	if err := ctx.BodyParser(&ids); err != nil {
		return err
	}

	res, err := c.bizSrv.Export(ctx.Context(), ids)
	if err != nil {
		return err
	}
	return ctx.Download(res, "导出.zip")
}

File: backend/src/main/java/pers/eastwind/billmanager/servicebill/controller/ServiceBillController.java
package pers.eastwind.billmanager.servicebill.controller;

import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import pers.eastwind.billmanager.attach.util.FileUtil;
import pers.eastwind.billmanager.common.model.ActionsResult;
import pers.eastwind.billmanager.common.model.PageResult;
import pers.eastwind.billmanager.servicebill.model.ServiceBillDTO;
import pers.eastwind.billmanager.servicebill.model.ServiceBillQueryParam;
import pers.eastwind.billmanager.servicebill.service.ServiceBillBizService;
import pers.eastwind.billmanager.servicebill.service.ServiceBillIOService;

import java.time.Instant;
import java.util.List;

/**
 * 服务单控制器
 */
@Slf4j
@RestController
@RequestMapping("/api/serviceBill")
@PreAuthorize("hasAnyRole('ADMIN', 'USER')")
public class ServiceBillController {
    private final ServiceBillBizService serviceBillBizService;
    private final ServiceBillIOService serviceBillIOService;

    public ServiceBillController(ServiceBillBizService serviceBillBizService, ServiceBillIOService serviceBillIOService) {
        this.serviceBillBizService = serviceBillBizService;
        this.serviceBillIOService = serviceBillIOService;
    }

    /**
     * 根据条件查找服务单
     *
     * @return 服务单列表
     */
    @PostMapping("/query")
    public PageResult<ServiceBillDTO> queryByParam(@RequestBody ServiceBillQueryParam queryParam) {

        Page<ServiceBillDTO> pageResult = serviceBillBizService.findByParam(queryParam);
        return PageResult.fromPage(pageResult);

    }

    /**
     * 根据 ID 获取服务单
     */
    @GetMapping("/{id}")
    public ServiceBillDTO getById(@PathVariable Integer id) {
        return serviceBillBizService.findById(id);
    }

    /**
     * 创建服务单
     *
     * @param serviceBillDTO 服务单
     * @return 保存后的服务单
     */
    @PostMapping
    public ServiceBillDTO create(@RequestBody ServiceBillDTO serviceBillDTO) {
        return serviceBillBizService.create(serviceBillDTO);
    }

    /**
     * 保存服务单
     *
     * @param serviceBillDTO 服务单
     * @return 保存后的服务单
     */
    @PutMapping
    public ServiceBillDTO save(@RequestBody ServiceBillDTO serviceBillDTO) {
        return serviceBillBizService.update(serviceBillDTO);
    }

    /**
     * 通过文件创建
     */
    @PostMapping("/import")
    public ServiceBillDTO importByFile(MultipartFile file)  {
        return serviceBillIOService.generateByFile(file.getResource());
    }

    /**
     * 删除服务单
     */
    @DeleteMapping
    public ActionsResult<Integer, Void> delete(@RequestBody List<Integer> ids) {
        return serviceBillBizService.delete(ids);
    }

    /**
     * 处理服务单
     */
    @PutMapping("/process")
    public ActionsResult<Integer, Void> process(@RequestBody List<Integer> ids) {
        return serviceBillBizService.process(ids);
    }

    /**
     * 处理完成服务单
     */
    @PutMapping("/processed")
    public ActionsResult<Integer, Void> processed(@RequestBody ProcessedParam param) {
        return serviceBillBizService.processed(param.ids, param.processedDate);
    }

    /**
     * 完成服务单
     */
    @PutMapping("/finish")
    public ActionsResult<Integer, Void> finish(@RequestBody FinishParam param) {
        return serviceBillBizService.finish(param.ids, param.finishedDate);
    }

    /**
     * 导出
     */
    @PostMapping(value = "/export", produces = "application/octet-stream")
    public Resource export(@RequestBody List<Integer> ids) {
        return FileUtil.loadByPath(serviceBillIOService.export(ids));
    }

    /**
     * 处理完成参数
     */
    public record ProcessedParam(List<Integer> ids, Instant processedDate) {
    }
    /**
     * 完成参数
     */
    public record FinishParam(List<Integer> ids, Instant finishedDate) {}
}
